### PR TITLE
bpo-33197: Update error message of invalid inspect.Parameters

### PR DESCRIPTION
**Pull request by corona10**
The error message from constructing invalid inspect.
Parameters was confused due to print integer enum value (parameter kind).
It is updated to print a string value of parameter kind.

https://bugs.python.org/issue33197

### CODE REVIEW

**Comment by corona10** [comment]
For reviewers,

Please add labels `needs backport to 3.7/3.6`

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
`kind` can be not only `_ParameterKind` member here, but an integer. You need first make a `_ParameterKind` from it: `_ParameterKind(kind)`.

There is the same formatting error at line 2465. The message at line 2739 can be improved too.

The message at line 2456 should contain the name of the type instead of the repr of the value.

**Comment by corona10** [comment]
@serhiy-storchaka
Thenk you for kind review, I will reflect your review soon.

**Comment by corona10** [comment]
I have made the requested changes; please review again

File: Lib/inspect.py [hunk 1]
```diff
@@ -2440,11 +2440,11 @@ def __init__(self, name, kind, *, default=_empty, annotation=_empty):
...
                         _VAR_POSITIONAL, _KEYWORD_ONLY, _VAR_KEYWORD):
             raise ValueError("invalid value for 'Parameter.kind' attribute")
-        self._kind = kind
+        self._kind = _ParameterKind(kind)
 
         if default is not _empty:
-            if kind in (_VAR_POSITIONAL, _VAR_KEYWORD):
-                msg = '{} parameters cannot have default values'.format(kind)
+            if self._kind in (_VAR_POSITIONAL, _VAR_KEYWORD):
+                msg = f'{self._kind!s} cannot have default values'
```
Comment (1st1) [inline Lib/inspect.py]: Why did you remove 'parameters' word from the message?

**Review by 1st1 — CHANGES_REQUESTED** [review]
This is a regression from blindly converting everything in stdlib to use enums. Thanks for fixing it!

I don't like the `'VAR_KEYWORD cannot have default values'` message, it's rather cryptic. Please add a mapping of parameter-kinds to human-readable form and use that mapping to render good-looking error messages. Thanks!

File: Lib/inspect.py (hunk 1 above)
Comment (corona10) [reply Lib/inspect.py]: @1st1 
No reason for it, I revert it.

**Comment by corona10** [comment]
@1st1 
I suggest using __str__ for the human-readable print and __repr__ for print enum key.
For example, in this case for `VAR_KEYWORD`
 __str__ ->` 'variadic keyword'`  __repr__ -> `'VAR_KEYWORD'`

What do you think?

**Comment by corona10** [comment]
@1st1 
Updated!

**Comment by corona10** [comment]
I have made the requested changes; please review again

File: Lib/inspect.py [hunk 2]
```diff
@@ -2453,19 +2461,18 @@ def __init__(self, name, kind, *, default=_empty, annotation=_empty):
...
             # See issue 19611.
             if kind != _POSITIONAL_OR_KEYWORD:
-                raise ValueError(
-                    'implicit arguments must be passed in as {}'.format(
-                        _POSITIONAL_OR_KEYWORD
-                    )
-                )
+                msg = 'implicit arguments must be passed in as {!s} arguments not {!s} arguments'
+                msg.format(_PARAM_NAME_MAPPING[_POSITIONAL_OR_KEYWORD],
+                           _PARAM_NAME_MAPPING[self._kind])
```
Comment (SylvainDe) [inline Lib/inspect.py]: Really naive question but would it better to have more consistency between line 2471 and 2474, ie have something like:

    if kind != _POS:
        msg.format(_PARAM[_POS], _PARAM[kind])

or:

    if self._kind != _POS:
        msg.format(_PARAM[_POS], _PARAM[self._kind])

In any case, I love the idea of adding more details in the error message :)

Comment (corona10) [reply Lib/inspect.py]: @SylvainDe Yeah, that might be better :)

**Comment by corona10** [comment]
@SylvainDe
Updated!

File: Lib/inspect.py [hunk 3]
```diff
@@ -2453,19 +2461,18 @@ def __init__(self, name, kind, *, default=_empty, annotation=_empty):
...
             # See issue 19611.
-            if kind != _POSITIONAL_OR_KEYWORD:
-                raise ValueError(
-                    'implicit arguments must be passed in as {}'.format(
-                        _POSITIONAL_OR_KEYWORD
-                    )
-                )
+            if self._kind != _POSITIONAL_OR_KEYWORD:
+                msg = 'implicit arguments must be passed in as {!s} arguments not {!s} arguments'
+                msg = msg.format(_PARAM_NAME_MAPPING[_POSITIONAL_OR_KEYWORD],
```
Comment (corona10) [inline Lib/inspect.py]: @1st1, @serhiy-storchaka 
Update nit fix for msg format and it was my mistake.
PTAL

**Comment by corona10** [comment]
@1st1
The review is delayed, If you have a time can you take a look please?

File: Lib/inspect.py [hunk 4]
```diff
@@ -2440,11 +2447,12 @@ def __init__(self, name, kind, *, default=_empty, annotation=_empty):
         if kind not in (_POSITIONAL_ONLY, _POSITIONAL_OR_KEYWORD,
```
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Just call `_ParameterKind(kind)` and catch an exception.

File: Lib/inspect.py [hunk 5]
```diff
@@ -2732,12 +2739,13 @@ def __init__(self, parameters=None, *, return_annotation=_empty,
                 kind_defaults = False
 
                 for idx, param in enumerate(parameters):
-                    kind = param.kind
+                    kind = _ParameterKind(param.kind)
```
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Is not `param.kind` always `_ParameterKind`?

**Comment by corona10** [comment]
@1st1, @serhiy-storchaka
Can you take a look?

File: Lib/inspect.py [hunk 6]
```diff
@@ -2440,11 +2447,12 @@ def __init__(self, name, kind, *, default=_empty, annotation=_empty):
...
             raise ValueError("invalid value for 'Parameter.kind' attribute")
-        self._kind = kind
+        self._kind = _ParameterKind(kind)
 
         if default is not _empty:
-            if kind in (_VAR_POSITIONAL, _VAR_KEYWORD):
-                msg = '{} parameters cannot have default values'.format(kind)
+            if self._kind in (_VAR_POSITIONAL, _VAR_KEYWORD):
+                msg = '{!s} parameters cannot have default values'
+                msg = msg.format(_PARAM_NAME_MAPPING[self._kind])
```
Comment (serhiy-storchaka) [inline Lib/inspect.py]: It may be convenient to add a helper for this. `_descr(self.kind)` looks better than `_PARAM_NAME_MAPPING[self.kind]`.

It can be defined as `_descr = _PARAM_NAME_MAPPING.__getitem__`.

File: Lib/inspect.py (hunk 4 above)
Comment (1st1) [reply Lib/inspect.py]: +1.

File: Lib/inspect.py (hunk 6 above)
Comment (1st1) [reply Lib/inspect.py]: I'd name the helper `_get_paramkind_descr()`

File: Lib/inspect.py (hunk 5 above)
Comment (1st1) [reply Lib/inspect.py]: Yes. I'd just add an `isinstance(param, Parameter)` if-check here.

File: Lib/inspect.py (hunk 6 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py]: The main benefit of the helper is that its name is shorter and you can write formatting in one line.

In 3.8 I would add an attribute to `_ParameterKind`, and this could be written as `self.kind.descr`. Or maybe override the `__str__` method and write it just as `self.kind`?

Comment (1st1) [reply Lib/inspect.py]: > The main benefit of the helper is that its name is shorter and you can write formatting in one line.

OTOH anything in Python can be callable and descriptive names with verbs in them make the code more readable.

> self.kind.descr

+1 for 3.8.  I'd call it `description` though :)  It a kind of API that you don't use frequently. But the code that uses introspection is usually complex, so it's nice when it's readable.

Comment (corona10) [reply Lib/inspect.py]: Thank you for all your kind review. I will update them tonight! Many thanks!

**Comment by corona10** [comment]
I have made the requested changes; please review again

**Comment by corona10** [comment]
@1st1 @serhiy-storchaka Please take a look

File: Lib/inspect.py [hunk 7]
```diff
@@ -2395,6 +2395,17 @@ class _ParameterKind(enum.IntEnum):
     def __str__(self):
         return self._name_
 
+    @property
+    def description(self):
+        param_name_mapping = {
```
Comment (1st1) [inline Lib/inspect.py]: I'd move this into a global variable to make it's faster.

Also, I want to backport this patch to 3.6 and 3.7.1 so maybe we should add the property in a separate PR that will only go to 3.8.

File: Lib/inspect.py [hunk 8]
```diff
@@ -2453,19 +2464,17 @@ def __init__(self, name, kind, *, default=_empty, annotation=_empty):
...
             # their name as "implicitN" and treat them as positional-only.
             # See issue 19611.
-            if kind != _POSITIONAL_OR_KEYWORD:
-                raise ValueError(
-                    'implicit arguments must be passed in as {}'.format(
-                        _POSITIONAL_OR_KEYWORD
-                    )
-                )
+            if self._kind != _POSITIONAL_OR_KEYWORD:
+                msg = 'implicit arguments must be passed in as {!s} arguments not {!s} arguments'
```
Comment (1st1) [inline Lib/inspect.py]: This line is longer than 79 characters (PEP8). Please reformat the code.

**Review by 1st1 — CHANGES_REQUESTED** [review]
Almost there. Please take a look at my comments.

File: Lib/inspect.py (hunk 7 above)
Comment (corona10) [reply Lib/inspect.py]: Okay, I will create a new PR which add a new property `description` for 3.8 after this PR is merged.

File: Lib/inspect.py [hunk 9]
```diff
@@ -2402,6 +2402,16 @@ def __str__(self):
 _KEYWORD_ONLY            = _ParameterKind.KEYWORD_ONLY
 _VAR_KEYWORD             = _ParameterKind.VAR_KEYWORD
 
+_PARAM_NAME_MAPPING = {
+    _POSITIONAL_ONLY: 'positional only',
+    _POSITIONAL_OR_KEYWORD: 'positional or keyword',
+    _VAR_POSITIONAL: 'variadic positional',
+    _KEYWORD_ONLY: 'keyword only',
+    _VAR_KEYWORD: 'variadic keyword'
```
Comment (1st1) [inline Lib/inspect.py]: "keyword-only"
"positional-only"

File: Lib/inspect.py [hunk 10]
```diff
@@ -2436,15 +2446,15 @@ class Parameter:
     empty = _empty
 
     def __init__(self, name, kind, *, default=_empty, annotation=_empty):
-
-        if kind not in (_POSITIONAL_ONLY, _POSITIONAL_OR_KEYWORD,
-                        _VAR_POSITIONAL, _KEYWORD_ONLY, _VAR_KEYWORD):
+        try:
+            self._kind = _ParameterKind(kind)
+        except Exception as e:
             raise ValueError("invalid value for 'Parameter.kind' attribute")
```
Comment (1st1) [inline Lib/inspect.py]:
```python
except ValueError:
    raise ValueError("invalid value for 'Parameter.kind' attribute") from None
```

File: Lib/inspect.py [hunk 11]
```diff
@@ -2736,8 +2746,9 @@ def __init__(self, parameters=None, *, return_annotation=_empty,
                     name = param.name
 
                     if kind < top_kind:
-                        msg = 'wrong parameter order: {!r} before {!r}'
-                        msg = msg.format(top_kind, kind)
+                        msg = 'wrong parameter order: {!s} parameter before {!s} parameter'
```
Comment (1st1) [inline Lib/inspect.py]: Long line. I suggest to add a ruler in your editor so that you can easily see when a line is longer than 79 chars.

File: Lib/inspect.py [hunk 12]
```diff
@@ -2453,19 +2463,19 @@ def __init__(self, name, kind, *, default=_empty, annotation=_empty):
...
             # See issue 19611.
-            if kind != _POSITIONAL_OR_KEYWORD:
-                raise ValueError(
-                    'implicit arguments must be passed in as {}'.format(
-                        _POSITIONAL_OR_KEYWORD
-                    )
-                )
+            if self._kind != _POSITIONAL_OR_KEYWORD:
+                msg = 'implicit arguments must be passed in as'
+                '{!s} arguments not {!s} arguments'
```
Comment (serhiy-storchaka) [inline Lib/inspect.py]: This line is ignored.

File: Lib/inspect.py [hunk 13]
```diff
@@ -2436,15 +2446,15 @@ class Parameter:
...
+            self._kind = _ParameterKind(kind)
+        except Exception as e:
             raise ValueError("invalid value for 'Parameter.kind' attribute")
-        self._kind = kind
 
         if default is not _empty:
-            if kind in (_VAR_POSITIONAL, _VAR_KEYWORD):
-                msg = '{} parameters cannot have default values'.format(kind)
+            if self._kind in (_VAR_POSITIONAL, _VAR_KEYWORD):
+                msg = '{!s} parameters cannot have default values'
```
Comment (serhiy-storchaka) [inline Lib/inspect.py]: `!s` is not needed since `_get_paramkind_descr()` returns a string.

File: Lib/inspect.py (hunk 10 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py]: But note that the second line is too long.

File: Lib/inspect.py (hunk 12 above)
Comment (corona10) [reply Lib/inspect.py]: Thank you for catch my mistake..

**Comment by corona10** [comment]
I have made the requested changes; please review again

File: Lib/inspect.py [hunk 14]
```diff
@@ -2436,15 +2446,16 @@ class Parameter:
...
 
     def __init__(self, name, kind, *, default=_empty, annotation=_empty):
-
-        if kind not in (_POSITIONAL_ONLY, _POSITIONAL_OR_KEYWORD,
-                        _VAR_POSITIONAL, _KEYWORD_ONLY, _VAR_KEYWORD):
-            raise ValueError("invalid value for 'Parameter.kind' attribute")
-        self._kind = kind
+        try:
+            self._kind = _ParameterKind(kind)
+        except ValueError as e:
```
Comment (serhiy-storchaka) [inline Lib/inspect.py]: `as e` is not needed.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM. Only one nitpick.

**Comment by corona10** [comment]
@1st1, @serhiy-storchaka: Updated!

**Comment by corona10** [comment]
@1st1 

Please take a look if you have a time 👍

File: Lib/test/test_inspect.py [hunk 15]
```diff
@@ -1444,6 +1444,19 @@ def f6(a, b, c):
         with self.assertRaisesRegex(TypeError, "'a', 'b' and 'c'"):
             inspect.getcallargs(f6)
 
+        # bpo-33197
+        with self.assertRaisesRegex(ValueError,
+                                    'variadic keyword parameters cannot have default values'):
```
Comment (1st1) [inline Lib/test/test_inspect.py]: Long line

File: Lib/test/test_inspect.py [hunk 16]
```diff
@@ -1444,6 +1444,19 @@ def f6(a, b, c):
...
+            inspect.Parameter("foo", kind=inspect.Parameter.VAR_KEYWORD,
+                              default=42)
+        with self.assertRaisesRegex(ValueError,
+                                    "invalid value for 'Parameter.kind' attribute"):
```
Comment (1st1) [inline Lib/test/test_inspect.py]: Long line

**Comment by corona10** [comment]
@1st1 

Thanks you for reviewing all 79 characters linting issues.
Now there are no longer characters over 79 characters on this patch
including inspect.py, test_inspect.py and news.

I should care about this linting convention from now on.

Thank you for redundunt reviewing for same issues :)

Enjoy your weekend! thank you

File: Lib/inspect.py [hunk 17]
```diff
@@ -2436,15 +2446,16 @@ class Parameter:
...
-
-        if kind not in (_POSITIONAL_ONLY, _POSITIONAL_OR_KEYWORD,
-                        _VAR_POSITIONAL, _KEYWORD_ONLY, _VAR_KEYWORD):
-            raise ValueError("invalid value for 'Parameter.kind' attribute")
-        self._kind = kind
+        try:
+            self._kind = _ParameterKind(kind)
+        except ValueError:
+            msg = "invalid value for 'Parameter.kind' attribute"
+            raise ValueError(msg) from None
```
Comment (1st1) [inline Lib/inspect.py]: last change here:

```python
raise ValueError(f'value {kind!r} is not a valid Parameter.kind')
```

File: Lib/inspect.py [hunk 18]
```diff
@@ -2453,19 +2464,22 @@ def __init__(self, name, kind, *, default=_empty, annotation=_empty):
...
             # See issue 19611.
-            if kind != _POSITIONAL_OR_KEYWORD:
-                raise ValueError(
-                    'implicit arguments must be passed in as {}'.format(
-                        _POSITIONAL_OR_KEYWORD
-                    )
+            if self._kind != _POSITIONAL_OR_KEYWORD:
+                msg = (
+                    'implicit arguments must be passed in as '
+                    '{} arguments not {} arguments'
```
Comment (1st1) [inline Lib/inspect.py]: "implicit arguments must be passed as positional or keyword arguments, not {}"

**Review by 1st1 — APPROVED** [review]
Left a couple nitpicks, please fix them. LGTM otherwise.

**Comment by corona10** [comment]
@1st1 @serhiy-storchaka 

Thanks! Updated!

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
Thanks, @corona10
